src/main/java/org/folio/rest/impl/InventoryHelper.java

```diff
@@ -0,0 +1,429 @@
+    // proceed with new Instance Record creation if no productId is provided
+    if (compPOL.getDetails() == null || compPOL.getDetails().getProductIds().isEmpty()) {
+      return createInstanceRecord(compPOL, productTypesMap);
+    }
+
+    String query = compPOL.getDetails().getProductIds().stream()
+      .map(productId -> buildProductIdQuery(productId, productTypesMap))
+      .collect(joining(" or "));
+
+    // query contains special characters so must be encoded before submitting
+    String endpoint = String.format(LOOKUP_INSTANCES_ENDPOINT, encodeQuery(query, logger), lang);
+
+    return handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger)
+      .thenCompose(instances -> {
+        if (instances.getJsonArray("instances").size() > 0) {
+          return completedFuture(extractId(getFirstObjectFromResponse(instances, "instances")));
+        }
+        return createInstanceRecord(compPOL, productTypesMap);
+      });
+  }
+
+  /**
+   * Creates Instance Record in Inventory and returns its Id.
+   *
+   * @param compPOL PO line to create Instance Record for
+   * @param productTypesMap product types Map used to build Instance Record json object
+   * @return id of newly created Instance Record
+   */
+  private CompletableFuture<String> createInstanceRecord(PoLine compPOL, Map<String, String> productTypesMap) {
+    JsonObject lookupObj = new JsonObject();
+    CompletableFuture<Void> instanceTypeFuture = getInstanceType(DEFAULT_INSTANCE_TYPE_CODE)
+      .thenAccept(lookupObj::mergeIn);
+    CompletableFuture<Void> statusFuture = getStatus(DEFAULT_STATUS_CODE)
+      .thenAccept(lookupObj::mergeIn);
+
+    return allOf(ctx, instanceTypeFuture, statusFuture)
+      .thenApply(v -> buildInstanceRecordJsonObject(compPOL, productTypesMap, lookupObj))
+      .thenCompose(instanceRecJson -> createRecordInStorage(instanceRecJson, String.format(CREATE_INSTANCE_ENDPOINT, lang)));
+  }
+
+  private CompletableFuture<JsonObject> getInstanceType(String typeName) {
+    String endpoint = String.format("/instance-types?query=code==%s", typeName);
+    return handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger);
+  }
+
+  private CompletableFuture<JsonObject> getStatus(String statusCode) {
+    String endpoint = String.format("/instance-statuses?query=code==%s", statusCode);
+    return handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger);
+  }
+
+  private String buildProductIdQuery(ProductId productId, Map<String, String> productTypes) {
+    return String.format("(identifiers adj \"\\\"identifierTypeId\\\": \\\"%s\\\"\" " +
+        "and identifiers adj \"\\\"value\\\": \\\"%s\\\"\")",
+      productTypes.get(productId.getProductIdType().toString()),
+      productId.getProductId());
+  }
+
+  private JsonObject buildInstanceRecordJsonObject(PoLine compPOL, Map<String, String> productTypes, JsonObject lookupObj) {
+    JsonObject instance = new JsonObject();
```

Comment:
It's a bummer we're forced to work with JsonObject here...  -- No action required --